File: Detect_Cycle.py
"""
Detect a cycle in a linked list. Note that the head pointer may be 'None' if the list is empty.

A Node is defined as: 
 
    class Node(object):
        def __init__(self, data = None, next_node = None):
            self.data = data
            self.next = next_node
"""


def has_cycle(head):
    slow = head
    fast = head

    while fast != None and fast.next != None:
        slow = slow.next  # Move slow pointer one step
        fast = fast.next.next  # Move fast pointer two steps

        # If there is a cycle, slow and fast pointers will meet
        if slow == fast:
            return True  # Cycle detected

    return False  # No cycle found

File: Detect_Cycle.js
function processData(input) {
  let slow = input;
  let fast = input;

  while (fast && fast.next) {
    slow = slow.next; // Move slow pointer one step
    fast = fast.next.next; // Move fast pointer two steps

    // If there is a cycle, slow and fast pointers will meet
    if (slow === fast) {
      return true; // Cycle detected
    }
  }
  return false; // No cycle found
}

process.stdin.resume();
process.stdin.setEncoding("ascii");
_input = "";
process.stdin.on("data", function (input) {
  _input += input;
});

process.stdin.on("end", function () {
  processData(_input);
});
